src/pybind/mgr/smart/module.py

```diff
@@ -1,33 +1,128 @@
+            "perm": "r"
+        },
+        {
+            "cmd": "osd smart scrape all",
+            "desc": "Scraping SMART data of all OSDs",
+            "perm": "r"
+        },
         {
             "cmd": "osd smart get "
                    "name=osd_id,type=CephString,req=true",
-            "desc": "Get smart data for osd.id",
+            "desc": "Get SMART data for osd_id",
+            "perm": "r"
+        },
+        {
+            "cmd": "osd smart get all",
+            "desc": "Get SMART data for all OSDs",
+            "perm": "r"
+        },
+        {
+            "cmd": "pool test",
+            "desc": "Just a pool test",
             "perm": "r"
         },
     ]
 
+    def open_connection(self):
+        # looking for the right pool
+        # TODO open pool, and if it fails - create and open
+        pools = self.rados.list_pools()
+        is_pool = False
+        for pool in pools:
+            self.log.debug('pool name is %s' % pool)
+            if pool == pool_name:
+                is_pool = True
+                break
+
+        # TODO make pool name configurable
+        if not is_pool:
+            self.rados.create_pool(pool_name)
+            self.log.debug('create %s pool' % pool_name)
+
+        ioctx = self.rados.open_ioctx(pool_name)
+
+        return (ioctx)
+
+    def handle_osd_smart_scrape_all(self):
+        osdmap = self.get("osd_map")
+        assert osdmap is not None
+
+        ioctx = self.open_connection()
+
+        for osd in osdmap['osds']:
+            osd_id = osd['osd']
+            self.do_osd_smart_scrape(str(osd_id), ioctx)
+
+        ioctx.close()
+        return (0, "", "")
+
+    def handle_osd_smart_scrape(self, osd_id):
+        ioctx = self.open_connection()
+        self.do_osd_smart_scrape(osd_id, ioctx)
+        ioctx.close()
+        return (0, "", "")
+
+    def do_osd_smart_scrape(self, osd_id, ioctx):
+        # running 'tell' command on OSD, retrieving smartctl results
+        result = CommandResult('')
+        self.send_command(result, 'osd', osd_id, json.dumps({
+            'prefix': 'smart',
+            'format': 'json',
+        }), '')
+        r, outb, outs = result.wait()
+        smart_data = json.loads(outb)
+
+        osd_host = self.get_metadata('osd', osd_id)['hostname']
+        self.log.debug('osd host %s' % osd_host)
+
+        for device, device_smart_data in smart_data.items():
+            object_name = osd_host + ":" + device
+            with rados.WriteOpCtx() as op:
+                # TODO should we change second arg to be byte?
+                # TODO what about the flag in ioctx.operate_write_op()?
+                now = str(time.time())
+                ioctx.set_omap(op, (now,), (str(json.dumps(device_smart_data)),))
+                ioctx.operate_write_op(op, object_name)
+                self.log.debug('writing object %s, key: %s, value: %s' % (object_name, now, str(json.dumps(device_smart_data))))
```

Comment:
can you confirm that it is writing key/value data with

rados -p smart_data ls -
rados -p smart_data listomapvals $object_name

I would make sure the date looks like "YYYYMMDD-HHMMSS".

Also, it might makes sense to have a *single* record for the entire smart scrape, and the value be the blob of json.

Reply:
1) can you confirm that it is writing key/value data:
- yes, here's a partial output:
https://pastebin.com/ds5CNuHP
all the examples I found used ioctx.set_omap() and ioctx.operate_write_op() to write objects, should I add op.create() regardless?

2) I would make sure the date looks like "YYYYMMDD-HHMMSS".
- done

3) Also, it might makes sense to have a single record for the entire smart scrape, and the value be the blob of json.
- not sure I understand: right now object names are $host:device, and for each object, omap key:val are $date:$json_blob_of_that_device. 
Should we add another object "$host" with omap key:val of $date:$json_blob_all_host_devices ?

Reply:
3. What you have is right--I just misread the code :)

Reply:
1. i would expect it to be needed or else the object won't get created.. but maybe i'm wrong?  try deleting the pool and doing a scrape and see if the object for the host:device gets recreated.